### ISSUE TITLE
cluster got into an unrecoverable state after deleting and recreating a collection

### ISSUE THREAD

**Issue by trashhalo** [qdrant/qdrant#1662]
## Current Behavior
Calls to, scroll, insert points, create collections and delete collections start to timeout. After a little bit all calls start 500ing with ` too many replicas returned an error` Errors from logs in detailed description.

## Steps to Reproduce
1. Have a couple collections loaded with 10k+ points each
2. Have 4 nodes in cluster configuration
3. Delete collections
4. Create collections with same name
5. Start reinserting points*
\* we insert about 25 points per call. concurrently between multiple application servers. 

## Expected Behavior
I should be able to delete and recreate indexes without consensus algorithm putting cluster into a unrecoverable state.

## Possible Solution
This isnt a fix but as a work around Im going to rewrite the index rebuild to create a new collection and cut over to it. Its a better implementation but was trying to avoid versioning collections.

How we recovered from this incident was uninstalling the helm application, removing the persistent volume claims, installing the helm application. Then updating running application server to point at the new empty cluster. Rebuild each collection one at a time. 😢 

## Context (Environment)
Our application workflow is primarily search based. The search cluster returning no data or timing out means that we had a site wide outage for about an hour.

Qdrant 1.1.0, 4x nodes in kube.
collections
shard 8
replication_factor 2

## Detailed Description
```
[2023-04-04T18:15:50.080Z WARN  storage::content_manager::consensus_manager] Failed to apply collection meta operation entry with user error: Wrong input: Replica 6850656706223262058 of shard 0 has state Some(Active), but expected Some(Initializing)

[2023-04-04T18:13:59.822Z WARN  storage::content_manager::consensus_manager] Failed to send message to http://qdrant-1.qdrant-headless:6335/ with error: Error in closure supplied to transport channel pool: status: DeadlineExceeded, message: "Timeout exceeded", details: [], metadata: MetadataMap { headers: {} }

[2023-04-04T18:11:10.294Z ERROR qdrant::consensus] Failed to propose entry: ProposalDropped

[2023-04-04T18:11:05.969Z WARN  collection::shards::replica_set] Failed to update shard COLLECTION_NAME:1 on peer 14108870475277683833, error: Service internal error: Tonic status error: The operation was cancelled

[2023-04-04T18:18:53.194Z WARN  qdrant::actix::helpers] error processing request: Waiting for consensus operation commit failed. Timeout set at: 10 seconds 
```

**Comment by trashhalo** [comment]
Discord thread https://discord.com/channels/907569970500743200/1092875487627464877

**Comment by trashhalo** [comment]
Id also settle for a way to kick a cluster into a consensus. Even if its just back to empty. Deleting the stateful set and persisted volumes is such a painful way to recover. It also extended the outage because I have to update the qdrant url on the deployed app to use the new cluster.

**Comment by DengYiping** [comment]
Bumping this. If we do it for a single collection, we also hit the same error here.

**Comment by agourlay** [comment]
I did some basic testing with concurrent creation and deletion of collection and I was unfortunately not able to reproduce this issue.

@DengYiping Can you please share as well some information about your setup please?

**Comment by timvisee** [comment]
I wonder if the underlying filesystem could be the problem.

Or if this is on macOS (I don't think so), we might flush incorrectly:

> For applications that require tighter guarantees about the integrity of
their data, Mac OS X provides the F_FULLFSYNC fcntl. The F_FULLFSYNC
fcntl asks the drive to flush all buffered data to permanent storage.
Applications, such as databases, that require a strict ordering of writes
should use F_FULLFSYNC to ensure that their data is written in the order
they expect. ([ref](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/fsync.2.html))

**Comment by DengYiping** [comment]
@timvisee 

I have the same issue again today. It seems like one single node corrupt the whole cluster. I was creating a index with 1 shard & replication factor of 1, and this issue appears. 

We are running in a 5 node setup on Kubernetes, backed by GP2 EBS volume.

This issue appears in one of the cluster health check test we implemented. This test does exactly 4 things using REST API:

- Creating a index called `collection_creation_test`
- Checking against "List collections" REST endpoint & see if the newly created index is there
- Checking against "Collection info" REST endpoint & see if there are any errors given -> this is where it fails
- Dropping the index called `collection_creation_test`

The above sequence is running okay for 1 week and all of a sudden on the weekend it started to fail.

Once it failed for the first time, there is no way to recover it unless:
 - Using a different collection name
 - Dropping the persistent volume & the data for the entire cluster

Probably worth replicating this issue using a script. Basically does the above 3 things & see if it will fail.

>  wonder if the underlying filesystem could be the problem.

We are using a GP2 volume on AWS, so it is very unlikely that the file system is the issue here.

**Comment by DengYiping** [comment]
I look a bit deeper into this. It seems like the whole cluster went into a rogue state. 

There are a few collections that doesn't even show up in every single node. Different nodes give different results in "get collections" endpoint.

**Comment by timvisee** [comment]
Thank you for getting back on this. It's a painful issue, and we should resolve it as soon as possible.

Could you share your cluster status for each node? And possibly the cluster status for each collection, on each node:

```bash
curl "http://NODE_URL/cluster" | jq
curl "http://NODE_URL/collections/COLLECTION_NAME/cluster" | jq
```

I wonder if there is a clear mismatch between commit/term numbers.

**Comment by generall** [comment]
Hi @DengYiping, can you please confirm that the cluster is running on the latest qdrant version (1.4.1 as of now)?

**Comment by DengYiping** [comment]
Yes, I can confirm that it is running on 1.4.1

**Comment by RafalGoslawski** [comment]
Additional info:

```bash
for PORT in 6333 6343 6353; do   curl -s "http://localhost:${PORT}/collections/test_collection/cluster" | jq -c .; done
```

Returns:

```json
{"result":{"peer_id":4799098437505445,"shard_count":3,"local_shards":[{"shard_id":1,"points_count":0,"state":"Active"}],"remote_shards":[{"shard_id":0,"peer_id":3413489478446639,"state":"Active"},{"shard_id":2,"peer_id":8458440906768741,"state":"Active"}],"shard_transfers":[]},"status":"ok","time":0.000100622}
{"result":{"peer_id":8458440906768741,"shard_count":3,"local_shards":[{"shard_id":2,"points_count":5,"state":"Active"}],"remote_shards":[{"shard_id":0,"peer_id":3413489478446639,"state":"Active"},{"shard_id":1,"peer_id":4799098437505445,"state":"Initializing"}],"shard_transfers":[]},"status":"ok","time":2.0166e-05}
{"result":{"peer_id":3413489478446639,"shard_count":3,"local_shards":[{"shard_id":0,"points_count":1,"state":"Active"}],"remote_shards":[{"shard_id":1,"peer_id":4799098437505445,"state":"Initializing"},{"shard_id":2,"peer_id":8458440906768741,"state":"Active"}],"shard_transfers":[]},"status":"ok","time":1.0833e-05}
```

**Comment by agourlay** [comment]
@RafalGoslawski Thank you very much for the fantastic work on providing a reproducer :clap: 

I have tried running it several times but so far the lagging node always recovers when the latency is reduced.

I will try a bit harder to make it work.

In the meantime, it would be fantastic if you could change the log level of "raft" to debug in your docker-compose file, re-run the test and then share with us the log files of each node for further analysis on our side.

Thanks again!

**Comment by RafalGoslawski** [comment]
Here's an interesting difference between the runs that reproduced the issue and the ones that didn't.

When the issue is reproduced, the nodes that we're not affected jump from `"term":2` to `"term":102` (on `node1` and `node2`) , so they presumably get the term data from the affected `node0`.

```
Running with 10000ms latency on node0
[...]
raft_status after 3 iterations:
Check raft
{"url":"http://localhost:6333","status":"ok","raft_info":{"term":95,"commit":141,"pending_operations":0,"leader":0,"role":"Candidate","is_voter":true}}
{"url":"http://localhost:6343","status":"ok","raft_info":{"term":2,"commit":152,"pending_operations":0,"leader":3010363724257094,"role":"Follower","is_voter":true}}
{"url":"http://localhost:6353","status":"ok","raft_info":{"term":2,"commit":152,"pending_operations":0,"leader":3010363724257094,"role":"Leader","is_voter":true}}
Running with 10ms latency on node0
[...]
raft_status after 1 iterations:
Check raft
{"url":"http://localhost:6333","status":"ok","raft_info":{"term":102,"commit":142,"pending_operations":0,"leader":3010363724257094,"role":"Follower","is_voter":true}}
{"url":"http://localhost:6343","status":"ok","raft_info":{"term":102,"commit":158,"pending_operations":0,"leader":3010363724257094,"role":"Follower","is_voter":true}}
{"url":"http://localhost:6353","status":"ok","raft_info":{"term":102,"commit":158,"pending_operations":0,"leader":3010363724257094,"role":"Leader","is_voter":true}}
Delete collection
```

When the issue is NOT reproduced, the affected node gets its term data from the non-affected nodes (`"term":1` to `"term":2` on `node0`).

```
Running with 10000ms latency on node0
[...]
raft_status after 3 iterations:
Check raft
{"url":"http://localhost:6333","status":"ok","raft_info":{"term":1,"commit":144,"pending_operations":0,"leader":2246616460745119,"role":"Leader","is_voter":true}}
{"url":"http://localhost:6343","status":"ok","raft_info":{"term":2,"commit":146,"pending_operations":0,"leader":4645553148523377,"role":"Follower","is_voter":true}}
{"url":"http://localhost:6353","status":"ok","raft_info":{"term":2,"commit":146,"pending_operations":0,"leader":4645553148523377,"role":"Leader","is_voter":true}}

Running with 10ms latency on node0
[...]
raft_status after 1 iterations:
Check raft
{"url":"http://localhost:6333","status":"ok","raft_info":{"term":2,"commit":148,"pending_operations":0,"leader":4645553148523377,"role":"Follower","is_voter":true}}
{"url":"http://localhost:6343","status":"ok","raft_info":{"term":2,"commit":148,"pending_operations":0,"leader":4645553148523377,"role":"Follower","is_voter":true}}
{"url":"http://localhost:6353","status":"ok","raft_info":{"term":2,"commit":148,"pending_operations":0,"leader":4645553148523377,"role":"Leader","is_voter":true}}
```

**Comment by agourlay** [comment]
Thanks!

Something weird I have noticed on my setup and now in your logs is that `raft=debug` does not seem to work with docker-compose logs.
Normally, every raft exchange generates a fair amount of structured logging (raft-rs uses slog).
That's really a bummer!

If you are still motivated to dig deeper, we could try having a look at the physical commit logs.
Could you please upload the raft data here as a tar archive for instance.
That is for each node, inside the storage directory:
- raft_state.json
- collection_meta_wal directory 

In the meantime I have been playing with your script with larger cluster but still no luck.
All instabilities eventually get resolve by the end of the script.

I must be missing something important.

**Comment by agourlay** [comment]
From your logs, it is indeed clearly visible that the lagging node is not catching up properly

```
node0  | [2023-08-24T14:45:49.900Z INFO  raft::raft] starting a new election, raft_id: 4826122942869202, term: 100
node0  | [2023-08-24T14:45:49.900Z INFO  raft::raft] became candidate at term 101, raft_id: 4826122942869202, term: 101
node0  | [2023-08-24T14:45:49.900Z INFO  raft::raft] broadcasting vote request, raft_id: 4826122942869202, to: [3010363724257094, 5293302778279663], log_index: 142, log_term: 1, term: 101, type: MsgRequestVote
node0  | [2023-08-24T14:45:49.900Z DEBUG qdrant::consensus] Changing hard state. New hard state: HardState { term: 101, vote: 4826122942869202, commit: 141 }
node0  | [2023-08-24T14:45:49.908Z DEBUG qdrant::consensus] Changing soft state. New soft state: SoftState { leader_id: 0, raft_state: Candidate }
node0  | [2023-08-24T14:45:49.934Z INFO  raft::raft] received votes response, raft_id: 4826122942869202, term: 101, type: MsgRequestVoteResponse, approvals: 1, rejections: 1, from: 5293302778279663, vote: false
node0  | [2023-08-24T14:45:49.934Z INFO  raft::raft] received votes response, raft_id: 4826122942869202, term: 101, type: MsgRequestVoteResponse, approvals: 1, rejections: 2, from: 3010363724257094, vote: false
node0  | [2023-08-24T14:45:49.935Z INFO  raft::raft] became follower at term 101, raft_id: 4826122942869202, term: 101
node0  | [2023-08-24T14:45:49.935Z DEBUG qdrant::consensus] Changing soft state. New soft state: SoftState { leader_id: 0, raft_state: Follower }
node0  | [2023-08-24T14:45:52.572Z INFO  raft::raft] received a message with higher term from 3010363724257094, raft_id: 4826122942869202, msg type: MsgRequestVote, message_term: 102, term: 101, from: 3010363724257094
node0  | [2023-08-24T14:45:52.572Z INFO  raft::raft] became follower at term 102, raft_id: 4826122942869202, term: 102
node0  | [2023-08-24T14:45:52.572Z INFO  raft::raft] [logterm: 1, index: 142, vote: 0] cast vote for 3010363724257094 [logterm: 2, index: 152] at term 102, raft_id: 4826122942869202, msg type: MsgRequestVote, term: 102, msg_index: 152, msg_term: 2, from: 3010363724257094, vote: 0, log_index: 142, log_term: 1
node0  | [2023-08-24T14:45:52.572Z DEBUG qdrant::consensus] Changing hard state. New hard state: HardState { term: 102, vote: 3010363724257094, commit: 141 }
node0  | [2023-08-24T14:45:52.599Z DEBUG qdrant::consensus] Changing soft state. New soft state: SoftState { leader_id: 3010363724257094, raft_state: Follower }
node0  | [2023-08-24T14:45:52.622Z INFO  raft::raft_log] found conflict at index 142, raft_id: 4826122942869202, conflicting term: 2, existing term: 1, index: 142
node0  | [2023-08-24T14:45:52.622Z DEBUG qdrant::consensus] Appending 1 entries to raft log
node0  | [2023-08-24T14:45:52.622Z DEBUG storage::content_manager::consensus::consensus_wal] Appending entry: Entry { entry_type: EntryNormal, term: 2, index: 142, data: [], context: [], sync_log: false }
```

1 - it starts an election
2 - gets a rejection
3 - becomes a follower
4 - receiving data conflicting due to difference in terms
5 - resolve the conflict and accepts the new entry for 142

At this point it should receive the remaining of the logs but it does not.

**Comment by RafalGoslawski** [comment]
I think I've ran into issues with Rust logs not reacting to env vars in docker before. I'll check if I can figure out how to get the debug logs from raft...
 
Here's the `/qdrant/storage` directories from all the nodes: [qdrant_raft_issue_1662.tar.gz](https://github.com/qdrant/qdrant/files/12437439/qdrant_raft_issue_1662.tar.gz)

**Comment by RafalGoslawski** [comment]
The missing debug logs might be because of `slog` disables trace and debug logs in release builds. See [notice](https://docs.rs/slog/latest/slog/#notable-details)

`Note: At compile time slog by default removes trace and debug level statements in release builds, and trace level records in debug builds. This makes trace and debug level logging records practically free, which should encourage using them freely. If you want to enable trace/debug messages or raise the compile time logging level limit, use the following in your Cargo.toml:`

```
slog = { version = ... ,
         features = ["max_level_trace", "release_max_level_warn"] }
```

**Comment by ffuugoo** [comment]
@RafalGoslawski Thanks for figuring out what's wrong with the logs. We've been trying to figure it out for the last hour or so. 🤦‍♀️

**Comment by agourlay** [comment]
@RafalGoslawski Great find :trophy: 

On my side, I made some progress thanks to your raft logs.

Using a [custom reader](https://github.com/qdrant/qdrant/pull/2244) I am able to see that the logs diverge on the last index for node0.

### node0

```
==========================
Entry ID:141
term:1
entry_type:0
data:"CollectionMeta(SetShardReplicaState(SetShardReplicaState { collection_name: \"test_collection\", shard_id: 0, peer_id: 3010363724257094, state: Active, from_state: Some(Initializing) }))"
==========================
Entry ID:142
term:1
entry_type:0
data:"CollectionMeta(DeleteCollection(DeleteCollectionOperation(\"test_collection\")))"
```

### node1

```
==========================
Entry ID:141
term:1
entry_type:0
data:"CollectionMeta(SetShardReplicaState(SetShardReplicaState { collection_name: \"test_collection\", shard_id: 0, peer_id: 3010363724257094, state: Active, from_state: Some(Initializing) }))"
==========================
Entry ID:142
term:2
entry_type:0
data:"[]"
==========================
Entry ID:143
term:2
entry_type:0
data:"CollectionMeta(CreateCollection(CreateCollectionOperation { collection_name: \"test_collection\", create_collection: CreateCollection { vectors: Single(VectorParams { size: 4, distance: Dot, hnsw_config: None, quantization_config: None, on_disk: None }), shard_number: Some(3), replication_factor: Some(1), write_consistency_factor: None, on_disk_payload: None, hnsw_config: None, wal_config: None, optimizers_config: None, init_from: None, quantization_config: None }, distribution: Some(ShardDistributionProposal { distribution: [(0, [3010363724257094]), (1, [4826122942869202]), (2, [5293302778279663])] }) }))"
==========================
```

### node2

```
==========================
Entry ID:141
term:1
entry_type:0
data:"CollectionMeta(SetShardReplicaState(SetShardReplicaState { collection_name: \"test_collection\", shard_id: 0, peer_id: 3010363724257094, state: Active, from_state: Some(Initializing) }))"
==========================
Entry ID:142
term:2
entry_type:0
data:"[]"
==========================
Entry ID:143
term:2
entry_type:0
data:"CollectionMeta(CreateCollection(CreateCollectionOperation { collection_name: \"test_collection\", create_collection: CreateCollection { vectors: Single(VectorParams { size: 4, distance: Dot, hnsw_config: None, quantization_config: None, on_disk: None }), shard_number: Some(3), replication_factor: Some(1), write_consistency_factor: None, on_disk_payload: None, hnsw_config: None, wal_config: None, optimizers_config: None, init_from: None, quantization_config: None }, distribution: Some(ShardDistributionProposal { distribution: [(0, [3010363724257094]), (1, [4826122942869202]), (2, [5293302778279663])] }) }))"
==========================
...
```

Although the conflict resolution for node0 says this it fixed it

```
node0  | [2023-08-24T14:45:52.622Z INFO  raft::raft_log] found conflict at index 142, raft_id: 4826122942869202, conflicting term: 2, existing term: 1, index: 142
node0  | [2023-08-24T14:45:52.622Z DEBUG qdrant::consensus] Appending 1 entries to raft log
node0  | [2023-08-24T14:45:52.622Z DEBUG storage::content_manager::consensus::consensus_wal] Appending entry: Entry { entry_type: EntryNormal, term: 2, index: 142, data: [], context: [], sync_log: false }
```

In practice it appears the log is not written to disk.
This is a great lead, we already had flush issue at this spot.

**Comment by RafalGoslawski** [comment]
I've built a local docker image with debug log level for raft and reproduced the issue again. Here's the log file: 
[docker-compose.raft-debug.log](https://github.com/qdrant/qdrant/files/12439366/docker-compose.raft-debug.log)

**Comment by RafalGoslawski** [comment]
There's a panic in the logs after the conflict was found:
```
node0  | 2023-08-25T12:53:53.927045Z DEBUG qdrant::consensus: Changing soft state. New soft state: SoftState { leader_id: 7610935524316430, raft_state: Follower }    
node0  | 2023-08-25T12:53:53.952437Z  INFO raft::raft_log: found conflict at index 149, raft_id: 4605021470587802, conflicting term: 2, existing term: 1, index: 149    
node0  | 2023-08-25T12:53:53.952544Z DEBUG raft::raft: Sending from 4605021470587802 to 7610935524316430, raft_id: 4605021470587802, msg: Message { msg_type: MsgAppendResponse, to: 7610935524316430
, from: 0, term: 0, log_term: 0, index: 149, entries: [], commit: 149, commit_term: 0, snapshot: None, request_snapshot: 0, reject: false, reject_hint: 0, context: [], deprecated_priority: 0, prior
ity: 0 }, to: 7610935524316430, from: 4605021470587802    
node0  | 2023-08-25T12:53:53.952610Z DEBUG qdrant::consensus: Appending 1 entries to raft log    
node0  | 2023-08-25T12:53:53.952662Z DEBUG storage::content_manager::consensus::consensus_wal: Appending entry: Entry { entry_type: EntryNormal, term: 2, index: 149, data: [], context: [], sync_log
: false }    
node0  | 2023-08-25T12:53:53.980351Z ERROR qdrant::startup: Panic backtrace: 
node0  |    0: std::backtrace_rs::backtrace::libunwind::trace
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
node0  |    1: std::backtrace_rs::backtrace::trace_unsynchronized
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
node0  |    2: std::backtrace::Backtrace::create
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/backtrace.rs:332:13
node0  |    3: qdrant::startup::setup_panic_hook::{{closure}}
node0  |              at ./src/startup.rs:19:25
node0  |    4: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1999:9
node0  |    5: std::panicking::rust_panic_with_hook
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:709:13
node0  |    6: std::panicking::begin_panic_handler::{{closure}}
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:595:13
node0  |    7: std::sys_common::backtrace::__rust_end_short_backtrace
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:151:18
node0  |    8: rust_begin_unwind
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:593:5
node0  |    9: core::panicking::panic_fmt
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/panicking.rs:67:14
node0  |   10: core::panicking::panic
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/panicking.rs:117:5
node0  |   11: storage::content_manager::consensus::consensus_wal::ConsensusOpWal::append_entries
node0  |              at ./lib/storage/src/content_manager/consensus/consensus_wal.rs:194:17
node0  |   12: storage::content_manager::consensus_manager::ConsensusManager<C>::append_entries
node0  |              at ./lib/storage/src/content_manager/consensus_manager.rs:656:9
node0  |   13: qdrant::consensus::Consensus::process_ready
node0  |              at ./src/consensus.rs:674:13
node0  |   14: qdrant::consensus::Consensus::on_ready
node0  |              at ./src/consensus.rs:618:39
node0  |   15: qdrant::consensus::Consensus::start
node0  |              at ./src/consensus.rs:457:34
node0  |   16: qdrant::consensus::Consensus::run::{{closure}}
node0  |              at ./src/consensus.rs:95:35
node0  |   17: std::sys_common::backtrace::__rust_begin_short_backtrace
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys_common/backtrace.rs:135:18
node0  |   18: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/thread/mod.rs:529:17
node0  |   19: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/panic/unwind_safe.rs:271:9
node0  |   20: std::panicking::try::do_call
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:500:40
node0  |   21: __rust_try
node0  |   22: std::panicking::try
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panicking.rs:464:19
node0  |   23: std::panic::catch_unwind
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/panic.rs:142:14
node0  |   24: std::thread::Builder::spawn_unchecked_::{{closure}}
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/thread/mod.rs:528:30
node0  |   25: core::ops::function::FnOnce::call_once{{vtable.shim}}
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/core/src/ops/function.rs:250:5
node0  |   26: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1985:9
node0  |   27: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/alloc/src/boxed.rs:1985:9
node0  |   28: std::sys::unix::thread::Thread::new::thread_start
node0  |              at /rustc/eb26296b556cef10fb713a38f3d16b9886080f26/library/std/src/sys/unix/thread.rs:108:17
node0  |   29: <unknown>
node0  |   30: <unknown>
node0  |     
node0  | 2023-08-25T12:53:53.980383Z ERROR qdrant::startup: Panic occurred in file lib/storage/src/content_manager/consensus/consensus_wal.rs at line 194: assertion failed: wal_index == index - offset    
node0  | 2023-08-25T12:53:54.050376Z DEBUG reqwest::connect: starting new connection: https://staging-telemetry.qdrant.io/    
node0  | 2023-08-25T12:53:54.288818Z ERROR qdrant::tonic::logging: gRPC /qdrant.Raft/Send unexpectedly failed with Internal error "Can't send Raft message over channel" 0.000426    
node0  | 2023-08-25T12:53:54.500904Z ERROR qdrant::tonic::logging: gRPC /qdrant.Raft/Send unexpectedly failed with Internal error "Can't send Raft message over channel" 0.000684    
node0  | 2023-08-25T12:53:54.712665Z ERROR qdrant::tonic::logging: gRPC /qdrant.Raft/Send unexpectedly failed with Internal error "Can't send Raft message over channel" 0.000466    
node0  | 2023-08-25T12:53:54.925402Z ERROR qdrant::tonic::logging: gRPC /qdrant.Raft/Send unexpectedly failed with Internal error "Can't send Raft message over channel" 0.000955    
node0  | 2023-08-25T12:53:55.131917Z ERROR qdrant::tonic::logging: gRPC /qdrant.Raft/Send unexpectedly failed with Internal error "Can't send Raft message over channel" 0.000723    
node0  | 2023-08-25T12:53:55.339734Z ERROR qdrant::tonic::logging: gRPC /qdrant.Raft/Send unexpectedly failed with Internal error "Can't send Raft message over channel" 0.000777   
```

**Comment by agourlay** [comment]
@RafalGoslawski I think I was able to fix it based on my previous observation.

I stumbled upon the same panic while writing my unit test :smile: 

https://github.com/qdrant/qdrant/pull/2508

**Comment by RafalGoslawski** [comment]
Ah, I've switched to the dev build for that one, so the debug_asserts kicked in.

I see there's a possible fix coming on the branch. I'll check if I can reproduce the issue with that branch.

**Comment by ffuugoo** [comment]
@agourlay @RafalGoslawski I've (finally) managed to reproduce the issue on current `dev`, and just re-run the test with the fix from #2508. There's a `found conflict at index ...` line in the logs, but the issue _did not_ reproduce with the fix.

**Comment by RafalGoslawski** [comment]
Ok, I've got 15 runs in a row without the issue. I think it's safe to assume the PR fixes the issue.

**Comment by RafalGoslawski** [comment]
I haven't found any other issues with more tests. 

Is there an ETA on when #2508 gets merged into master and released? This issue is a blocker for me, as it happens quite often in my setup.

**Comment by timvisee** [comment]
> I haven't found any other issues with more tests.

Thank you so much for going over this carefully. That's very helpful!

> Is there an ETA on when #2508 gets merged into master and released? This issue is a blocker for me, as it happens quite often in my setup.

We estimate to release a new version including the fix this Monday or Tuesday. Sorry for it taking some time.

**Comment by timvisee** [comment]
Hi @RafalGoslawski. It took us a bit longer, sorry about that.

But here's the good news, we've just released [v1.5.0](https://github.com/qdrant/qdrant/releases/tag/v1.5.0). It includes <https://github.com/qdrant/qdrant/pull/2508> which is what you're looking for. :smile: 

Since I believe this is resolved with this update, I'll close this now.
